README.md
Money library that takes care of dealing with double imprecision.


# Usage

```dart
import 'package:dart_price/dart_price.dart';
import 'package:intl/intl.dart';

void main() {
  final price = Price(1.2, CurrencyCode.EUR);
  final price2 = Price(1.1, CurrencyCode.EUR);
  print('add:');
  print(price + price2);
  print('substract:');
  print(price - price2);
  print('scale:');
  print(price * 2);
  print('divide:');
  print(price / 2);
  print('converted:');
  print(price.converted(CurrencyCode.USD, 1.1));
  print('sort:');
  print([price, price2]..sort());

  print('format:');
  final formatter = NumberFormat.currency(
    locale: 'fr',
    name: price.currency.code.name,
    symbol: price.currency.symbol,
  );
  print(formatter.format(price.amount.toDouble()));
}

```

CHANGELOG.md
## 0.0.3

- allow scalor to be int too 

## 0.0.2

- made currency map const

## 0.0.1

- Initial version.


test/dart_price_test.dart
import 'package:dart_price/dart_price.dart';
import 'package:test/test.dart';
import 'package:intl/intl.dart';

void main() {
  group('Price', () {
    test('Can create price with each currency', () {
      for (final code in CurrencyCode.values) {
        final price = Price(100, code);
        expect(price, isA<Price>());
      }
    });

    test('Can equal', () {
      expect(
        Price(100, CurrencyCode.USD),
        equals(Price(100, CurrencyCode.USD)),
      );
      expect(
        Price(100, CurrencyCode.USD),
        isNot(Price(101, CurrencyCode.USD)),
      );
      expect(
        Price(100, CurrencyCode.USD),
        isNot(Price(100, CurrencyCode.EUR)),
      );
    });

    test('Can be added', () {
      expect(
        Price(100, CurrencyCode.USD) + Price(100, CurrencyCode.USD),
        equals(
          Price(200, CurrencyCode.USD),
        ),
      );

      expect(
        Price(0.2, CurrencyCode.USD) + Price(0.1, CurrencyCode.USD),
        equals(
          Price(0.3, CurrencyCode.USD),
        ),
      );
    });

    test('Can be substracted', () {
      expect(
        Price(100, CurrencyCode.USD) - Price(200, CurrencyCode.USD),
        equals(
          Price(-100, CurrencyCode.USD),
        ),
      );

      expect(
        Price(0.2, CurrencyCode.USD) - Price(0.1, CurrencyCode.USD),
        equals(
          Price(0.1, CurrencyCode.USD),
        ),
      );
    });

    test('Can be scaled', () {
      expect(
        Price(100, CurrencyCode.USD) * 2,
        equals(
          Price(200, CurrencyCode.USD),
        ),
      );

      expect(
        Price(0.01, CurrencyCode.USD) * 100,
        equals(
          Price(1, CurrencyCode.USD),
        ),
      );
    });

    test('Can be sorted', () {
      final original = [
        Price(100, CurrencyCode.EUR),
        Price(1, CurrencyCode.EUR),
        Price(10, CurrencyCode.EUR),
      ];
      final sorted = [
        Price(1, CurrencyCode.EUR),
        Price(10, CurrencyCode.EUR),
        Price(100, CurrencyCode.EUR),
      ];
      expect(original..sort(), equals(sorted));
    });

    test('Can be serialized', () {
      final original = Price(100, CurrencyCode.EUR);
      final json = original.toJson();
      final deserialized = Price.fromJson(json);
      expect(original, deserialized);
    });

    test('Can be converted', () {
      final eur = Price(0.50, CurrencyCode.EUR);
      final usd = eur.converted(CurrencyCode.USD, 1.5);
      expect(usd, equals(Price(0.75, CurrencyCode.USD)));
    });

    test('Can be formatted locally', () {
      final price = Price(2000.50, CurrencyCode.EUR);
      final formatterEn = NumberFormat.currency(
        locale: 'en',
        name: price.currency.code.name,
        symbol: price.currency.symbol,
      );
      expect(formatterEn.format(price.amount.toDouble()), equals('€2,000.50'));
      final formatterFr = NumberFormat.currency(
        locale: 'fr',
        name: price.currency.code.name,
        symbol: price.currency.symbol,
      );
      expect(
          formatterFr
              .format(price.amount.toDouble())
              // remove space unicode character for easier testing
              .replaceAll(RegExp('\\s'), ''),
          equals('2000,50€'));
    });

    test('Can be generated from a locale string', () {
      final price = Price(2000.50, CurrencyCode.EUR);
      final formatterEn = NumberFormat.currency(
        locale: 'en',
        name: price.currency.code.name,
        symbol: price.currency.symbol,
      );
      final value = formatterEn.parse('€2,000.50');
      expect(Price(value.toDouble(), CurrencyCode.EUR), equals(price));
    });
  });
}


lib/src/price.dart
import 'dart:convert';

import 'package:decimal/decimal.dart';

import 'currency.dart';
import 'currency_code.enum.dart';

class Price implements Comparable<Price> {
  final Decimal amount;

  final Currency currency;

  const Price._(this.amount, this.currency);

  Price(double amount, CurrencyCode currencyCode)
      : amount = Decimal.parse(amount.toString()),
        currency = Currency.fromCode(currencyCode);

  Price abs() => Price._(amount.abs(), currency);

  Price converted(CurrencyCode currencyCode, double rate) => Price._(
        amount * Decimal.parse(rate.toString()),
        Currency.fromCode(currencyCode),
      );

  bool get isZero => amount == Decimal.zero;

  bool get isNegative => amount < Decimal.zero;

  /// Two prices are equal if their amount and currency are the same
  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;

    return other is Price &&
        other.amount == amount &&
        other.currency == currency;
  }

  /// Compares amounts of two [Price] instances.
  ///
  /// Throws exception if currencies do not equal (see [Currency.==]).
  bool operator >(Price other) {
    if (currency != other.currency) {
      _throwCurrencyError();
    }

    return amount > other.amount;
  }

  /// Compares amounts of two [Price] instances.
  ///
  /// Throws exception if currencies do not equal (see [Currency.==]).
  bool operator >=(Price other) {
    if (currency != other.currency) {
      _throwCurrencyError();
    }

    return amount >= other.amount;
  }

  /// Compares amounts of two [Price] instances.
  ///
  /// Throws exception if currencies do not equal (see [Currency.==]).
  bool operator <(Price other) {
    if (currency != other.currency) {
      _throwCurrencyError();
    }

    return amount < other.amount;
  }

  /// Compares amounts of two [Price] instances.
  ///
  /// Throws exception if currencies do not equal (see [Currency.==]).
  bool operator <=(Price other) {
    if (currency != other.currency) {
      _throwCurrencyError();
    }

    return amount <= other.amount;
  }

  /// Adds amount of the second [Price] operand and creates new instance.
  ///
  /// Throws exception if currencies do not equal (see [Currency.==]).
  Price operator +(Price other) {
    if (currency != other.currency) {
      _throwCurrencyError();
    }
    return Price._(amount + other.amount, currency);
  }

  /// Subtracts amount of the second [Price] operand and creates new instance.
  ///
  /// Throws exception if currencies do not equal (see [Currency.==]).
  Price operator -(Price other) {
    if (currency != other.currency) {
      _throwCurrencyError();
    }

    return Price._(amount - other.amount, currency);
  }

  Price operator *(num scalor) {
    return Price._(amount * Decimal.parse(scalor.toString()), currency);
  }

  Price operator /(num scalor) {
    final result = amount / Decimal.parse(scalor.toString());
    return Price._(result.toDecimal(), currency);
  }

  @override
  int compareTo(Price other) {
    if (this > other) return 1;
    if (this < other) return -1;
    return 0;
  }

  void _throwCurrencyError() {
    throw Exception('Operations with different currencies are not supported.');
  }

  @override
  int get hashCode => amount.hashCode ^ currency.hashCode;

  Map<String, dynamic> toMap() {
    return {
      'amount': amount.toJson(),
      'currency': currency.toJson(),
    };
  }

  factory Price.fromMap(Map<String, dynamic> map) {
    return Price._(
      Decimal.fromJson(map['amount']),
      Currency.fromJson(map['currency']),
    );
  }

  String toJson() => json.encode(toMap());

  factory Price.fromJson(String source) => Price.fromMap(json.decode(source));

  @override
  String toString() => 'Price(amount: $amount, currency: $currency)';
}


lib/src/currency.dart
import 'dart:convert';

import 'currency_code.enum.dart';

part 'currency_map.dart';

class Currency {
  /// The ISO-4217 currency code (e.g. USD).
  final CurrencyCode code;

  /// The ISO-4217 currency number (e.g. 840, for USD).
  final int number;

  /// The length of the fraction part of a currency.
  ///
  /// - for `134.23` it will be `2`
  /// - for `165` it will be `0`
  /// - for `111.2` it will be `1`
  final int digits;

  /// The currency name (e.g. US Dollar).
  final String name;

  /// The currency symbol (e.g. $).
  final String symbol;

  /// Constructs currency.
  ///
  /// Currency instance is immutable.
  const Currency._({
    required this.code,
    required this.number,
    required this.digits,
    required this.name,
    required this.symbol,
  });

  /// Finds currency by ISO-4217 currency code (e.g. USD).
  factory Currency.fromCode(
    /// ISO-4217 code to look for.
    CurrencyCode code,
  ) =>
      currencyMap[code]!;

  /// Finds currency by ISO-4217 currency code (e.g. USD) as a string.
  factory Currency.fromString(
    /// ISO-4217 code to look for.
    String codeString,
  ) {
    final code = CurrencyCode.values.byName(codeString.toUpperCase());
    return currencyMap[code]!;
  }

  @override
  String toString() {
    return 'Currency(code: $code, number: $number, digits: $digits, name: $name, symbol: $symbol)';
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;

    return other is Currency &&
        other.code == code &&
        other.number == number &&
        other.digits == digits &&
        other.name == name &&
        other.symbol == symbol;
  }

  @override
  int get hashCode {
    return code.hashCode ^
        number.hashCode ^
        digits.hashCode ^
        name.hashCode ^
        symbol.hashCode;
  }

  Map<String, dynamic> toMap() {
    return {
      'code': code.name,
    };
  }

  factory Currency.fromMap(Map<String, dynamic> map) {
    return Currency.fromCode(CurrencyCode.values.byName(map['code']));
  }

  String toJson() => json.encode(toMap());

  factory Currency.fromJson(String source) =>
      Currency.fromMap(json.decode(source));
}


lib/src/currency_code.enum.dart
// ignore_for_file: constant_identifier_names

enum CurrencyCode {
  AED,
  AFN,
  ALL,
  AMD,
  ANG,
  AOA,
  ARS,
  AUD,
  AWG,
  AZN,
  BAM,
  BBD,
  BDT,
  BGN,
  BIF,
  BMD,
  BND,
  BOB,
  BRL,
  BSD,
  BWP,
  BZD,
  CAD,
  CDF,
  CHF,
  CLP,
  CNY,
  COP,
  CRC,
  CVE,
  CZK,
  DJF,
  DKK,
  DOP,
  DZD,
  EGP,
  ETB,
  EUR,
  FJD,
  FKP,
  GBP,
  GEL,
  GIP,
  GMD,
  GNF,
  GTQ,
  GYD,
  HKD,
  HNL,
  HRK,
  HTG,
  HUF,
  ILS,
  INR,
  ISK,
  JMD,
  JPY,
  KES,
  KGS,
  KHR,
  KMF,
  KRW,
  KYD,
  KZT,
  LAK,
  LBP,
  LKR,
  LRD,
  LSL,
  MAD,
  MDL,
  MGA,
  MKD,
  MMK,
  MNT,
  MOP,
  MUR,
  MVR,
  MWK,
  MXN,
  MYR,
  MZN,
  NAD,
  NGN,
  NIO,
  NOK,
  NPR,
  NZD,
  PAB,
  PEN,
  PGK,
  PHP,
  PKR,
  PLN,
  PYG,
  QAR,
  RON,
  RSD,
  RUB,
  RWF,
  SAR,
  SBD,
  SCR,
  SEK,
  SGD,
  SLL,
  SOS,
  SRD,
  SVC,
  SZL,
  THB,
  TJS,
  TOP,
  TRY,
  TTD,
  TWD,
  TZS,
  UAH,
  UGX,
  USD,
  UYU,
  UZS,
  VND,
  VUV,
  WST,
  XAF,
  XCD,
  XOF,
  XPF,
  YER,
  ZAR,
  ZMW,
  BHD,
  BTN,
  CUP,
  ERN,
  IDR,
  IRR,
  IQD,
  JOD,
  KPW,
  KWD,
  LYD,
  OMR,
  SHP,
  SSP,
  SYP,
  TND,
  VES,
  MRU,
  STN,
  CUC,
  ZWL,
  BYN,
  TMT,
  GHS,
  SDG
}


example/dart_price_example.dart
import 'package:dart_price/dart_price.dart';
import 'package:intl/intl.dart';

void main() {
  final price = Price(1.2, CurrencyCode.EUR);
  final price2 = Price(1.1, CurrencyCode.EUR);

  print('format:');
  print(buildEnglishFormatter(price).format(price.amount.toDouble()));
  print(buildFrenchFormatter(price).format(price.amount.toDouble()));
  print('add:');
  print(price + price2);
  print('substract:');
  print(price - price2);
  print('scale:');
  print(price * 2);
  print('divide:');
  print(price / 2);
  print('converted:');
  print(price.converted(CurrencyCode.USD, 1.1));
  print('sort:');
  print([price, price2]..sort());
}

/// the formatters are from the intl library but prices work well with it.
NumberFormat buildFrenchFormatter(Price price) {
  return NumberFormat.currency(
    locale: 'fr',
    name: price.currency.code.name,
    symbol: price.currency.symbol,
  );
}

NumberFormat buildEnglishFormatter(Price price) {
  return NumberFormat.currency(
    locale: 'en',
    name: price.currency.code.name,
    symbol: price.currency.symbol,
  );
}
